### PR TITLE
[SHELLEXT][MYDOCS][INF] Add mydocs.dll and .mydocs file extension

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

I found "SendTo My Documents" is realized by `.mydocs` filename extension in WinXP/2k3. To implement this, I will add `mydocs.dll` file.

mydocs

JIRA issue: [CORE-12562](https://jira.reactos.org/browse/CORE-12562)

AFTER:
after

### CODE REVIEW

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    WCHAR szSrc[MAX_PATH];
+    for (UINT iItem = 0; iItem < cItems; ++iItem)
+    {
+        DragQueryFileW(hDrop, iItem, szSrc, MAX_PATH);
+
+        if (!PathFileExistsW(szSrc))
+        {
+            *pdwEffect = 0;
+            DragLeave();
+            return E_FAIL;
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:92]: Might be nice to indicate which file failed.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        if (iItem > 0)
+            strSrcList += L"|";
+        strSrcList = szSrc;
+    }
+    strSrcList += L"|";
+    strSrcList += L"|";
+
+    for (INT i = 0, cch = strSrcList.GetLength(); i < cch; ++i)
+    {
+        // because strSrcList[i] is constant, we have to do workaround...
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:104]: No!
This is a hack, not a workaround.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        if (strSrcList[i] == L'|')
+            memcpy(const_cast<WCHAR *>(&strSrcList[i]), L"\0", sizeof(WCHAR));
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:106]: The proper way is:
```
LPWSTR Ptr = strSrcList.GetBuffer();
// Do your modifications
strSrcList.ReleaseBuffer();
```
This way you tell the `CString` that you are going to modify it, and if this is a shared string (which is possible!) that it should create a copy you can safely modify.
With the `ReleaseBuffer` you say that you are doing modifying it, and that it is under control of the CString again.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 2 above)
Comment (learn-more) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:104]: In general, when something is const, that is for a reason!

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        return E_FAIL;
+    }
+
+    // "My Documents"
+    WCHAR szDir[MAX_PATH + 1];
+    SHGetSpecialFolderPathW(NULL, szDir, CSIDL_PERSONAL, FALSE);
+    szDir[lstrlenW(szDir) + 1] = 0;
+
+    HDROP hDrop = reinterpret_cast<HDROP>(stg.hGlobal);
+    UINT cItems = ::DragQueryFileW(hDrop, 0xFFFFFFFF, NULL, 0);
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:80]: I am not a big fan of extracting info this way, but I am not sure we have enough to do it better :(

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp]: OK. Thanks.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp]: Thanks. Hack is removed.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:96]: Error message is added.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:80]: It seems like our `IDataObject` `HDROP` doesn't support multiple file items...

Comment (learn-more) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:80]: Yeah, we should probably write a helper for this, that tries a few methods to extract files.
First using HIDA, then falling back to legacy systems like HDROP...

Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp]: I used HIDA. Done.

Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp]: I used HIDA. Done.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    // get the path of "My Documents"
+    WCHAR szzDir[MAX_PATH + 1];
+    SHGetSpecialFolderPathW(NULL, szzDir, CSIDL_PERSONAL, FALSE);
+    szzDir[lstrlenW(szzDir) + 1] = 0; // ends with double NULs
+
+    // for all source items
+    CStringW strSrcList;
+    WCHAR szSrc[MAX_PATH];
+    const BYTE *pb = reinterpret_cast<BYTE *>(pida);
+    PCIDLIST_ABSOLUTE pidlParent = reinterpret_cast<PCIDLIST_ABSOLUTE>(pb + pida->aoffset[0]);
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:99]: `HIDA_GetPIDLFolder`

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    for (iItem = 0; iItem < cItems; ++iItem)
+    {
+        // query source pidl
+        PCITEMID_CHILD pidlChild = reinterpret_cast<PCITEMID_CHILD>(pb + pida->aoffset[iItem + 1]);
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:103]: `HIDA_GetPIDLItem`

Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:103]: Fixed in https://github.com/reactos/reactos/commit/2dc2ec9de359e4d6c31b6f61ba305e4323db8ddd.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:99]: Fixed in https://github.com/reactos/reactos/commit/2dc2ec9de359e4d6c31b6f61ba305e4323db8ddd.

**Pull request merged**